src/solver/softmax/attn_softmax.cpp

```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/softmax/solvers.hpp>
+
+#include <miopen/softmax/invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/softmax.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/target_properties.hpp>
+#include <miopen/float_equal.hpp>
+
+MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_DEBUG_ATTN_SOFTMAX)
+
+namespace miopen::solver::softmax {
+
+namespace {
+template <typename T, typename S = std::enable_if_t<std::is_unsigned_v<T>, T>>
+constexpr inline S Ceil(const T val, const T div)
+{
+    return (val - 1 + div) / div;
+}
+
+constexpr uint32_t nextPow2(uint32_t v)
```

Comment:
We already have nextPow2 function implemented in regular softmax solver. Could you please reuse that code ?

Reply:
Exactly, I just wanted to highlight the problem.

We already had lots of implementations of the same or similar functions function across the project.
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/forwardNaive.cpp#L54
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/include/miopen/pooling.hpp#L39
more generic "nextPow2":
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/conv_winoRxS.cpp#L106
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/backward2d.cpp#L113
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/forward2d.cpp#L86
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/forwardNd.cpp#L73

As well as integer ceil division:
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/conv_winoRxS.cpp#L100
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/conv_winoRxS_fused.cpp#L50
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/batchnorm/problem_description.cpp#L171
and around 300 other places with just an inline version of this function, for example:
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/backwardNd.cpp#L112.
https://github.com/ROCm/MIOpen/blob/efaafebc51c4db6f0f2731df7e2e366f5b0952fd/src/solver/pooling/forwardNd.cpp#L144

And there is no place for such kind of helper functions in the project. @atamazov probably you have any ideas?

Reply:
I think it has to be fixed properly - move all the identical functions not just from here, but from all the project, into a dedicated helper header, well document the header ant let everyone read it.

Reply:
Issue created https://github.com/ROCm/MIOpen/issues/2748

Meanwhile `nextPow2` from the regular softmax is not mathematically correct - it should not be a signed integer. It works for positive values but does not make sense for the negative (and even it triggers `signed integer` overflow, which is UB).
Unsigned version makes more sense and may trigger `unsigned integer` overflow which is perfectly defined.